pkg/registry/core/namespace/storage/storage.go

```diff
@@ -185,16 +185,38 @@ func (r *REST) Delete(ctx context.Context, name string, options *metav1.DeleteOp
 					existingNamespace.Status.Phase = api.NamespaceTerminating
 				}
 
-				// Remove orphan finalizer if options.OrphanDependents = false.
-				if options.OrphanDependents != nil && *options.OrphanDependents == false {
-					// remove Orphan finalizer.
-					newFinalizers := []string{}
-					for i := range existingNamespace.ObjectMeta.Finalizers {
-						finalizer := existingNamespace.ObjectMeta.Finalizers[i]
-						if string(finalizer) != metav1.FinalizerOrphanDependents {
-							newFinalizers = append(newFinalizers, finalizer)
-						}
+				// remove orphan and foregroundDeletion first, add them back if needed
```

Comment:
Why is the logic structured this way? I am finding it very hard to follow. I would expect:

```
// 1. collect data
currentFinalizers := map[string]bool{}
for range existingNamespace.ObjectMeta.Finalizers { ... }

// 2. compute which finalizers it should have
shouldHaveFinalizers := map[string]bool{
  metav1.FinalizerOrphanDependents: shouldHaveOrphanFinalizer(options, ...),
  metav1.FinalizerDeleteDependents: shouldHaveForgroundFinalizer(options, ...),
}

// 3. do we need to make changes?
for k, v := range shouldHaveFinalizers {
  if v2, ok := currentFinalizers[k]; !ok || v != v2 {
    changeNeeded = true; break
  }
}

// 4. Make changes
if changeNeeded {
  ...
}
```

Putting the "shouldHaveXXXFinalizer" logic in a separate function will make it easier to unit test.

Reply:
Ok, let me organize the code in the suggested way.

Reply:
Thanks, sorry for being a pain!

Reply:
Updated. Thanks! (That's a great suggestion!)

Reply:
Thanks, much easier to follow now. (Next time we change this I'll recommend not mutating the "current" map, but this is fine for now. :) )